tests/http_spec.rs

```diff
@@ -0,0 +1,216 @@
+    let mut response = Response { status: status_code, ..Default::default() };
+
+    // Insert headers from the mock into the response.
+    for (key, value) in mock_response.0.headers {
+      let header_name = HeaderName::from_str(&key)?;
+      let header_value = HeaderValue::from_str(&value)?;
+      response.headers.insert(header_name, header_value);
+    }
+
+    // Set the body of the response.
+    response.body = ConstValue::try_from(serde_json::from_value::<Value>(mock_response.0.body)?)?;
+
+    Ok(response)
+  }
+}
+
+#[tokio::test]
+async fn test_body_yaml() {
+  let spec = HttpSpec::read("tests/data/sample.yaml").unwrap();
+  assert_downstream(spec).await;
+}
+
+async fn assert_downstream(spec: HttpSpec) {
+  let has_only_annotation = spec
+    .downstream_assertions
+    .iter()
+    .any(|assertion| matches!(assertion.annotation, Some(Annotation::Only)));
+
+  for downstream_assertion in spec.downstream_assertions.iter() {
+    match &downstream_assertion.annotation {
+      Some(Annotation::Skip) if !has_only_annotation => {
+        let request_details = format_request_details(&downstream_assertion.request.0);
+        println!("Skipping test in: {}\nRequest Details: {}", spec.name, request_details);
+        continue;
+      }
+      Some(Annotation::Only) | None
+        if !has_only_annotation || matches!(downstream_assertion.annotation, Some(Annotation::Only)) =>
+      {
+        let response = run(spec.clone(), &downstream_assertion).await.unwrap();
+        let body = hyper::body::to_bytes(response.into_body()).await.unwrap();
+        assert_eq!(
+          body,
+          serde_json::to_string(&downstream_assertion.response.0.body).unwrap()
+        );
+      }
+      Some(Annotation::Fail) if !has_only_annotation => {
+        let response = run(spec.clone(), &downstream_assertion).await;
+        assert!(response.is_err());
+      }
+      _ => {} // Skip other cases if "Only" is present in any assertion
+    }
+  }
+}
+
+// Helper function to format request details for printing.
+fn format_request_details(request: &APIRequest) -> String {
+  format!(
+    "Method: {:?}, Path: {}, Body: {}",
+    request.method.clone(),
+    request.url.clone(),
+    request.body.clone()
+  )
+}
+
+#[tokio::test]
+async fn test_body_json() {
+  let spec = HttpSpec::read("tests/data/sample.json").unwrap();
+  assert_downstream(spec).await;
+}
+
+async fn run(spec: HttpSpec, downstream_assertion: &&DownstreamAssertion) -> anyhow::Result<hyper::Response<Body>> {
+  let query_string = serde_json::to_string(&downstream_assertion.request.0.body).expect("body is required");
+  let method = downstream_assertion.request.0.method.clone();
+  let url = downstream_assertion.request.0.url.clone();
+  let state = spec.setup().await;
+  let req = Request::builder()
+    .method(method)
+    .uri(url.as_str())
+    .body(Body::from(query_string));
+  graphql_request(req?, state.as_ref()).await
```

Comment:
- [ ] Log `ok` after each file is executed. Like we do it in GraphQLSpec.
- [ ] Make sure tests for each file run in parallel.

Reply:
Done